std-bits/table/src/main/java/org/enso/table/write/DelimitedWriter.java

```diff
@@ -0,0 +1,191 @@
+    this.writeHeaders = writeHeaders;
+    emptyValue = this.quote + "" + this.quote;
+  }
+
+  public void write(Table table) throws IOException {
+    int numberOfColumns = table.getColumns().length;
+    assert numberOfColumns == columnFormatters.length;
+
+    if (writeHeaders) {
+      boolean quoteAllHeaders = writeQuoteBehavior == WriteQuoteBehavior.ALWAYS;
+      for (int col = 0; col < numberOfColumns; ++col) {
+        boolean isLast = col == numberOfColumns - 1;
+        writeCell(table.getColumns()[col].getName(), isLast, quoteAllHeaders);
+      }
+    }
+
+    int numberOfRows = table.rowCount();
+    for (int row = 0; row < numberOfRows; ++row) {
+      for (int col = 0; col < numberOfColumns; ++col) {
+        boolean isLast = col == numberOfColumns - 1;
+        Object cellValue = table.getColumns()[col].getStorage().getItemBoxed(row);
+        String formatted = columnFormatters[col].format(cellValue);
+        boolean wantsQuoting =
+            writeQuoteBehavior == WriteQuoteBehavior.ALWAYS && wantsQuotesInAlwaysMode(cellValue);
+        writeCell(formatted, isLast, wantsQuoting);
+      }
+    }
+
+    output.flush();
+  }
+
+  private boolean wantsQuotesInAlwaysMode(Object value) {
+    return value instanceof String || !isNonTextPrimitive(value);
+  }
+
+  private boolean isNonTextPrimitive(Object value) {
+    return value instanceof Long
+        || value instanceof Double
+        || value instanceof Boolean
+        || value instanceof LocalDate
+        || value instanceof LocalDateTime
+        || value instanceof LocalTime
+        || value instanceof ZonedDateTime;
+  }
+
+  private boolean quotingEnabled() {
+    return writeQuoteBehavior != WriteQuoteBehavior.NEVER;
+  }
+
+  private void writeCell(String value, boolean isLastInRow, boolean wantsQuoting)
+      throws IOException {
+    String processed = value == null ? "" : quotingEnabled() ? quote(value, wantsQuoting) : value;
+    output.write(processed);
+    if (isLastInRow) {
+      output.write(NEWLINE);
+    } else {
+      output.write(delimiter);
+    }
+  }
+
+  private boolean needsQuoting(String value) {
+    boolean containsDelimiter = value.indexOf(delimiter) >= 0;
+    boolean containsQuote = value.indexOf(quoteChar) >= 0;
+    boolean containsQuoteEscape = value.indexOf(quoteEscapeChar) >= 0;
+
+    return containsDelimiter || containsQuote || containsQuoteEscape;
+  }
+
+  /**
+   * Wraps the value in quotes, escaping any characters if necessary.
+   *
+   * <p>The {@code wantsQuoting} parameter allows to request quoting even if it wouldn't normally be
+   * necessary. This is used to implement the `always_quote` mode for text and custom objects.
+   */
+  private String quote(String value, boolean wantsQuoting) {
+    if (value.isEmpty()) {
+      return emptyValue;
+    }
+
+    boolean shouldQuote = wantsQuoting || needsQuoting(value);
+    if (!shouldQuote) {
+      return value;
+    }
+
+    String escaped;
```

Comment:
Fold the contains booleans into here.

```suggestion
    String escaped = value;
    if (containsQuoteEscape) {
      escaped = escaped.replace(quoteEscape, quoteEscapeReplacement);
    }
    if (containsQuote) {
      escaped = escaped.replace(quote, quoteReplacement);
    }
```
(for this block)